lib/runtime-core/src/backing.rs

```diff
@@ -103,6 +103,104 @@ impl LocalBacking {
         })
     }
 
+    /// Resets the `LocalBacking` (`Memories` and `Globals`) for an `Instance` using the provided `ModuleInfo`
+    pub(crate) fn reset(&self, module_info: &ModuleInfo) -> RuntimeResult<()> {
+        Self::reset_memories(&module_info, &self.memories)?;
+        Self::reset_globals(&module_info, &self.globals)
+    }
+
+    fn reset_memories(
+        module_info: &ModuleInfo,
+        memories: &SliceMap<LocalMemoryIndex, Memory>,
+    ) -> RuntimeResult<()> {
+        Self::shrink_memories(memories)?;
+        Self::zero_memories(memories);
+        Self::reinitialize_memories(module_info, memories)
+    }
```

Comment:
```suggestion
    fn reset_memories(
        module_info: &ModuleInfo,
        memories: &SliceMap<LocalMemoryIndex, Memory>,
    ) -> RuntimeResult<()> {
        // zero the previously used memory
        Self::zero_memories(memories);

        Self::shrink_memories(memories)?;
        Self::reinitialize_memories(module_info, memories)
    }
```

The old pages should be zeroed as well (as in, first call `zero_memories` and only after that call `shrink_memories`). Otherwise, when obtaining new pages they will contain the data from the last execution instead of being zeroed.
Relevant notes:
- [grow](https://github.com/ElrondNetwork/wasmer/blob/1347511dd2fda59fad75c5df56c11a4859ec397c/lib/runtime-core/src/memory/static_.rs#L65) does not zero out the memory, it simply assumes that it already is zeroed
- when allocating static pages, the call which allocates the memory [Memory::with_size](https://github.com/ElrondNetwork/wasmer/blob/1347511dd2fda59fad75c5df56c11a4859ec397c/lib/runtime-core/src/sys/unix/memory.rs#L110) calls libc::mmap, which [according to the docs](https://www.man7.org/linux/man-pages/man2/mmap.2.html) zeroes out the memory (because the `MAP_ANON` argument is provided)

Reply:
Its a good catch, but I don't think it applies here, as we use handle `dynamic memory` and `unshared`
Just for sanity, I did it anyway, no cost involved.